core/lib/node/src/resource/lazy_resource.rs

```diff
@@ -0,0 +1,84 @@
+use std::sync::{Arc, Mutex};
+
+use thiserror::Error;
+use tokio::sync::watch;
+
+use super::Resource;
+use crate::node::StopReceiver;
+
+/// A lazy resource represent a resource that isn't available at the time when the tasks start.
+/// Normally it's used to represent the resources that should be provided by one task to another one.
+/// Lazy resources are aware of the node lifecycle, so attempt to resolve the resource won't hang
+/// if the resource is never provided: the resolve future will fail once the stop signal is sent by the node.
+#[derive(Debug)]
+pub struct LazyResource<T: Resource> {
+    resource: Arc<Mutex<Option<T>>>,
+    resolve_receiver: watch::Receiver<bool>,
+    resolve_sender: Arc<Mutex<watch::Sender<bool>>>,
```

Comment:
why Mutex on watch? Watch access is atomic anyway (I think it has a mutex internally).